src/main/java/com/commercetools/sync/sdk2/products/utils/AttributeUtils.java

```diff
@@ -1,95 +1,28 @@
-      return AttributeAccessor.asSetReference(attribute);
-    }
-    return Collections.emptyList();
-  }
-
-  /**
-   * Given an attribute this method first checks if its of type "nested" or "set" of "nested" and
-   * transforms its value into a {@link List} of {@link Attribute}. The resulting list is then
-   * mapped to type "reference" or "set" of "reference" depending on attributes' values types.
-   *
-   * @param attribute - Attribute to extract the Reference if exists
-   * @return - a {@link List} of {@link Reference} extracted from the given attribute or empty list
-   *     if the attribute doesn't contain reference types.
-   */
-  public static List<Reference> getAttributeReferences(@Nonnull final Attribute attribute) {
-    final List<Reference> referenceList = getAttributeReference(attribute);
-    if (!referenceList.isEmpty()) {
-      return referenceList;
-    } else {
-      final List<Attribute> flattenedAttributes = flatMapNestedAttributes(attribute);
-      return flattenedAttributes.stream()
-          .map(AttributeUtils::getAttributeReference)
-          .flatMap(Collection::stream)
-          .collect(Collectors.toList());
-    }
-  }
-
-  /**
-   * Takes an object and checks if it's of type {@link List} of {@link Attribute} which is
-   * significant for a nested attribute.
-   *
-   * @param attrValue - value to be checked
-   * @return true, if the given param is of type {@link List} of {@link Attribute} otherwise false
-   */
-  private static boolean isNested(@Nonnull final Object attrValue) {
-    return attrValue instanceof List
-        && (((List) attrValue).stream().allMatch(v -> v instanceof Attribute));
-  }
-
-  /**
-   * Given an attribute this method first checks if its of type "nested" or "set" of "nested" and
-   * transforms its value into a flattened {@link List} of {@link Attribute} using recursion.
-   *
-   * @param attribute - Attribute of any value
-   * @return - a {@link List} of {@link Attribute} transformed from the given attribute.
-   */
-  private static List<Attribute> flatMapNestedAttributes(@Nonnull final Attribute attribute) {
-    final Object attrValue = attribute.getValue();
-    List<Attribute> flattenedAttributes;
-    if (isNested(attrValue)) {
-      flattenedAttributes = AttributeAccessor.asNested(attribute);
-    } else if (attrValue instanceof List
-        && ((List) attrValue).stream().allMatch(v -> isNested(v))) {
-      final List<List<Attribute>> nestedSet = AttributeAccessor.asSetNested(attribute);
-      flattenedAttributes =
-          nestedSet.stream().flatMap(Collection::stream).collect(Collectors.toList());
-    } else {
-      return singletonList(attribute);
-    }
-    return flattenedAttributes.stream()
-        .map(attr -> flatMapNestedAttributes(attr))
-        .flatMap(Collection::stream)
-        .collect(Collectors.toList());
+  @Nonnull
+  public static List<JsonNode> getAttributeReferencesAsJson(@Nonnull final Attribute attribute) {
```

Comment:
So returning Reference was not a correct way?

Reply:
I've splitted the method to make it more clear what's done: [7308732](https://github.com/commercetools/commercetools-sync-java/pull/1088/commits/73087329061ab9452135ae77b863b42240389082)
The converted value has to be set in the attribute to be able to replace the references in async streams.

Reply:
Maybe I didn't ask clearly. I was wondering why Reference is not used anymore? What is the problem that you changed everywhere to JsonNode instead?

Reply:
The problem is, that as attribute value is typed with Object. Coming from different sources, it can be Map, JsonNode, String, Reference ... To overcome the complexity checkin all those different types i've changed to JsonNode.